File: web/tests/reference/tries/implement-trie.py
class Trie:
    def __init__(self):
        self.root = {}

    def insert(self, word):
        node = self.root
        for ch in word:
            node = node.setdefault(ch, {})
        node["$"] = True

    def _walk(self, s):
        node = self.root
        for ch in s:
            if ch not in node:
                return None
            node = node[ch]
        return node

    def search(self, word):
        node = self._walk(word)
        return node is not None and "$" in node

    def starts_with(self, prefix):
        return self._walk(prefix) is not None

File: web/tests/reference/tries/implement-trie.js
class Trie {
  constructor() {
    this.root = new Map();
  }

  insert(word) {
    let node = this.root;
    for (const ch of word) {
      if (!node.has(ch)) node.set(ch, new Map());
      node = node.get(ch);
    }
    node.set("$", true);
  }

  walk(s) {
    let node = this.root;
    for (const ch of s) {
      if (!node.has(ch)) return null;
      node = node.get(ch);
    }
    return node;
  }

  search(word) {
    const node = this.walk(word);
    return node !== null && node.has("$");
  }

  startsWith(prefix) {
    return this.walk(prefix) !== null;
  }
}
